planning.py

```diff
@@ -294,6 +329,207 @@ def add_vertices(
+        """Use two trees, one starting from start one from goal, to explore the space"""
+        start_time = time.time()
+        ta = RRT(self.workspace, start)
+        tb = RRT(self.workspace, goal)
+        v_ta = ta.graph.add_vertex(start)
+        v_tb = tb.graph.add_vertex(goal)
+        for _ in range(k):
+            # plt.figure()
+            # ax = plt.gca()
+            # self.workspace.draw(ax)
+            # self.draw(ax)
+            # ax.plot(start[0], start[1], "o", color="g")
+            # ax.plot(goal[0], goal[1], "o", color="r")
+            # ta.draw(ax,rgb=(1,0,0))
+            # tb.draw(ax,rgb=(0,1,0))
+
+            # plt.show()
+            qn = self.workspace.sample()
+            v_nearest, dist = ta.closest_vertex(qn)
+            if dist < min_edge_len:
+                continue
+            q_nearest = v_nearest.coord
+
+            if dist > max_edge_len:
+                qn = q_nearest + (qn - q_nearest) * max_edge_len / dist
+
+            if ta.workspace.edge_is_in_collision(q_nearest, qn):
+                continue
+            if self.workspace.point_is_in_collision(q):
+                q = self.closest_point_not_in_collision(q, q_nearest)
+            vn = ta.graph.add_vertex(qn)
+            ta.graph.add_edge(vn, v_nearest)
+
+            # operate on second tree
+            qn_prime = self.workspace.sample()
+            v_nearest, dist = tb.closest_vertex(qn_prime)
+            if dist < min_edge_len:
+                continue
+            q_nearest = v_nearest.coord
+
+            if dist > max_edge_len:
+                qn_prime = q_nearest + (qn_prime - q_nearest) * max_edge_len / dist
+
+            if tb.workspace.edge_is_in_collision(q_nearest, qn_prime):
+                continue
+            if self.workspace.point_is_in_collision(q):
+                q = self.closest_point_not_in_collision(q, q_nearest)
+            vn_prime = tb.graph.add_vertex(qn_prime)
+            tb.graph.add_edge(vn_prime, v_nearest)
+
+            if self.touch(ta, tb):
+                self.preprocessing_time = 0
+                self.query_time = time.time() - start_time
+                self.ta = ta
+                self.tb = tb
+                return
+            if ta.graph.n - tb.graph.n > 10:
+                ta, tb = tb, ta
+        self.preprocessing_time = 0
+        self.query_time = time.time() - start_time
+        self.ta = None
+        self.tb = None
+
+    def touch(self, ta, tb, goal_dist=0.5):
+        """determine if two trees are touching each other"""
+        v, v_tb, dist = ta.closest_vertex_graph(tb)
+        if dist < goal_dist:
+            v.connect(v_tb)
+            ta.graph.add_edge(v, v_tb)
+            return True
+        return False
+
+    def draw(self, ax):
+        if self.ta is None:
+            print("No solution")
+        else:
+            self.ta.draw(ax, rgb=(1, 0, 0))
+            self.tb.draw(ax, rgb=(0, 1, 0))
+
+
+class RRT_further_reachin(RRT):
```

Comment:
I'd also like to call this something like `UnboundedRRT`, but not exactly sure on the name.